### PR TITLE
[Impeller] dont emulate command buffers for compute on Metal/Vulkan.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Part of the work for: https://github.com/flutter/flutter/issues/140804

Removes ComputeCommand and has both ComputePassMTL and ComputePassVK immediately record to their respective comand buffers. This combines the commits for Metal/Vulkan as they are both fairly small. After this, I can add barriers to fix compute for Vulkan.

### CODE REVIEW

File: impeller/renderer/backend/metal/compute_pass_bindings_cache_mtl.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+//-----------------------------------------------------------------------------
+/// @brief      Ensures that bindings on the pass are not redundantly set or
+///             updated. Avoids making the driver do additional checks and makes
+///             the frame insights during profiling and instrumentation not
+///             complain about the same.
+///
+///             There should be no change to rendering if this caching was
+///             absent.
+///
+struct ComputePassBindingsCache {
```
Comment (chinmaygarde) [inline impeller/renderer/backend/metal/compute_pass_bindings_cache_mtl.h:25]: `ComputePassBindingsCacheMTL`. The MTL suffix for non-backend-agnostic classes.

File: impeller/renderer/backend/metal/compute_pass_mtl.h [hunk 2]
```diff
@@ -9,9 +9,96 @@
...
+ private:
+  struct BufferOffsetPair {
+    id<MTLBuffer> buffer = nullptr;
+    size_t offset = 0u;
+  };
+  using BufferMap = std::map<uint64_t, BufferOffsetPair>;
+  using TextureMap = std::map<uint64_t, id<MTLTexture>>;
+  using SamplerMap = std::map<uint64_t, id<MTLSamplerState>>;
+
+  id<MTLComputeCommandEncoder> encoder_;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/metal/compute_pass_mtl.h:95]: `= nil` (or `nullptr` like the rest) this out. IIRC, ObjC ivars aren't zero initialized in C++.

File: impeller/renderer/backend/metal/compute_pass_mtl.mm [hunk 3]
```diff
@@ -42,231 +47,113 @@
...
-                *data.texture.resource)) {
-        return false;
-      }
+  // TODO(dnfield): use feature detection to support non-uniform threadgroup
+  // sizes.
+  // https://github.com/flutter/flutter/issues/110619
+  auto width = grid_size.width;
+  auto height = grid_size.height;
+
+  auto maxTotalThreadsPerThreadgroup = static_cast<int64_t>(
```
Comment (chinmaygarde) [inline impeller/renderer/backend/metal/compute_pass_mtl.mm:124]: Here and elsewhere, the use `underscore_case` for vars instead of `camelCase`.

Comment (chinmaygarde) [reply impeller/renderer/backend/metal/compute_pass_mtl.mm:124]: Oh, I see this was just code that was moved. Just the same, a drive-by change to make this conform the style would be good.

File: impeller/renderer/backend/metal/compute_pass_mtl.h (hunk 2 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/metal/compute_pass_mtl.h]: Done

**Pull request merged**